eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/EventCriteriaBuilderTagStage.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2010-2025. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.eventsourcing.eventstore;
+
+import jakarta.annotation.Nonnull;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Interface declaring the possible builder actions for the tag stage of the {@link EventCriteriaBuilder}. After calling
+ * any of the methods defined in this interface, the builder will complete and return an {@link EventCriteria}.
+ *
+ * @author Mitchell Herrijgers
+ * @see EventCriteria
+ * @see EventCriteriaBuilder
+ * @since 5.0.0
+ */
+public interface EventCriteriaBuilderTagStage {
+
+    /**
+     * Define that the event must contain all the provided {@code tags} to match. These tags function in an AND
+     * relation, meaning that an event must have all tags to match. A partial match is not sufficient.
+     *
+     * @param tags The tags to match against.
+     * @return The completed {@link EventCriteria} instance.
+     */
+    EventCriteria withTags(@Nonnull Set<Tag> tags);
```

Comment:
Looking at the `EventCriteriaTest` just now (really love the additions btw!), I was wondering if it would make sense to rename the methods of this second stage to `andTags`. Thus, you would get a flow like this:

```java
// sample one
EventCriteria.matches().anyEventTypes().andAnyTags();
// sample two
EventCriteria.matches().eventTypes("type").andAnyTags();
// sample three
EventCriteria.matches().eventTypes("type").andTags("id","value");
```